src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py

```diff
@@ -306,6 +315,42 @@ def validate(self, node, options=None):
                     "transformation can only be applied to the outermost "
                     "Range.")
 
+        if not options:
+            options = {}
+        allow_string_array = options.get("allow_string", False)
+        # If we allow string arrays then we can skip the check.
+        if not allow_string_array:
+            lhs = assignment.lhs
+            # ArrayMixin datatype lookup can fail if the indices contain a
+            # Call or Intrinsic Call. We catch this exception and continue
+            # for now - TODO #1799
+            try:
+                if lhs.datatype.intrinsic == ScalarType.Intrinsic.CHARACTER:
+                    raise TransformationError(
+                        "The NemoArrayRange2LoopTrans transformation doesn't "
+                        "allow character arrays by default. This can be "
+                        "enabled by passing the allow_string option to the "
+                        "transformation."
+                    )
+            except NotImplementedError:
+                pass
+            for child in assignment.rhs.walk((Literal, Reference)):
+                try:
+                    # Skip unresolved types
+                    if isinstance(child.datatype,
+                                  (UnresolvedType, UnsupportedType)):
+                        continue
+                    if (child.datatype.intrinsic ==
+                            ScalarType.Intrinsic.CHARACTER):
+                        raise TransformationError(
+                            "The NemoArrayRange2LoopTrans transformation "
+                            "doesn't allow character arrays by default. This "
+                            "can be enabled by passing the allow_string "
+                            "option to the transformation."
+                        )
```

Comment:
Nemo still fails at line 328 with:
```
psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py", line 328, in validate
    if lhs.datatype.intrinsic == ScalarType.Intrinsic.CHARACTER:
       ^^^^^^^^^^^^^^^^^^^^^^
    AttributeError: 'UnsupportedFortranType' object has no attribute 'intrinsic'
```
This is because that check is not preceded with the:
```
# Skip unresolved types
if isinstance(child.datatype, (UnresolvedType, UnsupportedType)):
    continue
```
as the right-hand-side check. If both need to be the same, could we merge both together by doing the loop for `for child in assignment.walk((Literal, Reference)):` and remove the top try?

Then, I see we let strings inside UnsupportedTypes pass the validation. This is fine (for now) but could you mention that in the docstring and include again the `character(LEN=100)`-style test. I would like to have what happens in this case clarified in docstring and tests because it is common in NEMO.